![img](/Users/hongviet/Documents/GitHub/Data-Analysis/thuchanh/BTVN/cau1.jpg)

## 🎯 Confusion Matrix
- True Positive (TP): 80  
- False Positive (FP): 20  
- True Negative (TN): 90  
- False Negative (FN): 10  

---

## 🧮 Các công thức đo lường hiệu suất

### 1. **Accuracy (Độ chính xác)**
- ![img](/Users/hongviet/Documents/GitHub/Data-Analysis/thuchanh/BTVN/acc.png)

- ➡️ Accuracy = (80 + 90) / (80 + 90 + 20 + 10) = 170 / 200 = 0.85 (85%)
- __**ÝNghĩa**__:
    + Accuracy cho biết tỷ lệ dự đoán đúng trên toàn bộ dữ liệu.
    
    + Đây là một độ đo tổng quát nhưng không phù hợp khi dữ liệu mất cân bằng 
    + (ví dụ: 90% là âm tính, 10% là dương tính).
### 2. **Precision (Độ chính xác dương tính)**
- ![img](/Users/hongviet/Documents/GitHub/Data-Analysis/thuchanh/BTVN/pre.png)
- ➡️ Precision = 80 / (80 + 20) = 80 / 100 = 0.80 (80%)
- **Ý nghĩa**:
    + Precision đo lường tỷ lệ các mẫu được dự đoán là dương tính thực sự là dương tính.
    + Thích hợp khi chi phí của sai dương tính (False Positive) là cao.
    + Ví dụ: chẩn đoán ung thư, xét nghiệm sai gây lo lắng và điều trị không cần thiết.
### 3. **Recall (Độ nhạy / True Positive Rate)**
- Recall = TP / (TP + FN)
- ➡️ Recall = 80 / (80 + 10) = 80 / 90 ≈ 0.8889 (88.89%)
- Ý nghĩa:
    + Recall cho biết trong tất cả các trường hợp dương tính thực sự, mô hình phát hiện được bao nhiêu.
    + Quan trọng khi không được bỏ sót các trường hợp dương tính (ví dụ: phát hiện bệnh, gian lận tài chính).
### 4. **F1-score**
- ![img](./f1.png)
- ➡️ F1 = 2 × (0.80 × 0.8889) / (0.80 + 0.8889) ≈ 2 × 0.7111 / 1.6889 ≈ 0.841
- Ý nghĩa:
    + F1 là một độ đo cân bằng giữa precision và recall.
    + Nó đặc biệt hữu ích khi cần cân bằng giữa sai dương tính và sai âm tính.
    + F1 đạt giá trị cao nhất khi precision và recall đều cao.
### 5. **Specificity (True Negative Rate)**
- Specificity = TN / (TN + FP)
- ➡️ Specificity = 90 / (90 + 20) = 90 / 110 ≈ 0.8182 (81.82%)
- Ý nghĩa:
    + Specificity cho biết trong số các trường hợp âm tính thực sự, có bao nhiêu được dự đoán đúng.
    + Rất quan trọng trong các ứng dụng cần đảm bảo không gây dương tính giả 
    + (ví dụ: xét nghiệm tiền sản, kiểm soát chất lượng).

## ✅ Tổng kết

| Độ đo       | Giá trị     |
|-------------|-------------|
| Accuracy    | 85.00%      |
| Precision   | 80.00%      |
| Recall      | 88.89%      |
| F1-score    | 84.10%      |
| Specificity | 81.82%      |


![img](/Users/hongviet/Documents/GitHub/Data-Analysis/thuchanh/BTVN/cau2.jpg)

In [ ]:
# bảng dữ liệu
#giá trị thực tế
y_true = [3.0, 4.5, 5.0 , 6.0, 3.5]
#giá trị dự đoán
y_pred = [2.5, 4.0, 5.2, 5.5, 3.0]

#tính toán sai số
# trung bình tuyệt đối
def mean_absolute_error(y_true, y_pred):
    n = len(y_true)
    total_error = 0
    for i in range(n):
        total_error += abs(y_true[i] - y_pred[i])
    return total_error / n

#căn bật trung bình bình phương RMSE
def root_mean_squared_error(y_true, y_pred):
    n = len(y_true)
    total_error = 0
    for i in range(n):
        total_error += (y_true[i] - y_pred[i]) ** 2
    return (total_error / n) ** 0.5

# tính R2
def r_squared(y_true, y_pred):
    n = len(y_true)
    ss_total = 0
    ss_residual = 0
    mean_y_true = sum(y_true) / n
    for i in range(n):
        ss_total += (y_true[i] - mean_y_true) ** 2
        ss_residual += (y_true[i] - y_pred[i]) ** 2
    return 1 - (ss_residual / ss_total)

# tính toán các chỉ số
mae = mean_absolute_error(y_true, y_pred)
rmse = root_mean_squared_error(y_true, y_pred)
r2 = r_squared(y_true, y_pred)
# in kết quả
print("Mean Absolute Error (MAE):", mae)
print("Root Mean Squared Error (RMSE):", rmse)
print("R-squared (R2):", r2)


Mean Absolute Error (MAE): 0.44000000000000006
Root Mean Squared Error (RMSE): 0.4560701700396552
R-squared (R2): 0.8175438596491228


MAE = 0.44 triệu VND, tức sai số trung bình khoảng 440 nghìn đồng, là khá nhỏ.

RMSE = 0.456 triệu VND xác nhận rằng sai số nghiêm trọng không nhiều.

R² = 0.8175 (≈ 81.75%) là mức giải thích khá tốt, tức mô hình giải thích được ~82% phương sai trong dữ liệu.

➡️ Kết luận:
Mô hình dự đoán tương đối chính xác, sai số thấp và độ phù hợp tốt. Có thể cải thiện thêm bằng cách thử mô hình phi tuyến, thêm biến hoặc tinh chỉnh tham số.

![img](/Users/hongviet/Documents/GitHub/Data-Analysis/thuchanh/BTVN/cau3.jpg)

 -  __a)__ mô hình phù hợp để phát hiện đúng càng nhiều bệnh nhân thật sự là ***M3*** bởi vì tỷ lệ độ nhạy của mô hình này là tốt nhất, phù hợp với việc nhận diện sự bất thường trong nhận biết bệnh nhân và bất thường trong gia dịch 
 - __b)__ nếu muốn chọn mô hình cân bằng giữa __precision__ và __recall__ thì ta nên chọn mô hình M3 vì đây là mô hình có chỉ số ***F1*** cao nhất

![img](/Users/hongviet/Documents/GitHub/Data-Analysis/thuchanh/BTVN/cau4.jpg)

    a)Giải thích khả năng xảy ra tình trạng này

### Tình trạng này rất có thể xảy ra do tập dữ liệu bị mất cân bằng (imbalanced dataset).

####     - Accuracy (Độ chính xác) đo lường tỷ lệ tổng số dự đoán đúng trên toàn bộ tập dữ liệu ((TP + TN) / Tổng). Khi dữ liệu mất cân bằng, lớp đa số (majority class) chiếm tỷ lệ rất lớn. Nếu mô hình chỉ đơn giản dự đoán hầu hết các điểm dữ liệu thuộc về lớp đa số, nó vẫn có thể đạt được Accuracy cao, vì số lượng dự đoán đúng cho lớp đa số (TN) là rất lớn.
####     - F1-score là trung bình điều hòa (harmonic mean) của Precision và Recall. Nó nhạy cảm hơn với hiệu suất của mô hình trên lớp thiểu số (minority class).
####         + Precision (TP / (TP + FP)): Độ chuẩn xác - Tỷ lệ các điểm được dự đoán là dương tính mà thực sự là dương tính.
####         + Recall (TP / (TP + FN)): Độ phủ - Tỷ lệ các điểm thực sự dương tính mà mô hình dự đoán đúng.
####     - Trong trường hợp dữ liệu mất cân bằng, mô hình có Accuracy cao nhưng F1-score thấp thường là do nó dự đoán kém trên lớp thiểu số. Nó có thể có Recall thấp (bỏ lỡ nhiều điểm thuộc lớp thiểu số - nhiều FN) hoặc Precision thấp (dự đoán sai nhiều điểm thành lớp thiểu số - nhiều FP), hoặc cả hai. Điều này kéo F1-score xuống thấp, mặc dù Accuracy vẫn cao do dự đoán đúng phần lớn các điểm thuộc lớp đa số.

    b) Nếu tập dữ liệu bị mấy cân bằng bạn sẽ áp dụng các kỹ thuật nào để cải thiện mô hình? Liệt kê ít nhất 3 kỹ thuật

Có nhiều kỹ thuật để cải thiện mô hình khi gặp dữ liệu mất cân bằng. Dưới đây là ít nhất 3 kỹ thuật phổ biến:

1. Thay đổi tỷ lệ mẫu (Resampling):
   -  Oversampling (Lấy mẫu quá mức): Tăng số lượng mẫu của lớp thiểu số. Ví dụ: Lặp lại ngẫu nhiên các mẫu thiểu số, hoặc tạo ra các mẫu tổng hợp mới (ví dụ: kỹ thuật SMOTE - Synthetic Minority Over-sampling Technique).
   -  Undersampling (Lấy mẫu dưới mức): Giảm số lượng mẫu của lớp đa số. Ví dụ: Loại bỏ ngẫu nhiên các mẫu đa số, hoặc sử dụng các kỹ thuật chọn lọc hơn (ví dụ: Tomek links, NearMiss).
   -  Kết hợp Oversampling và Undersampling.
2. Sử dụng các độ đo đánh giá khác: Thay vì chỉ dựa vào Accuracy, tập trung vào các độ đo phù hợp hơn cho dữ liệu mất cân bằng như Precision, Recall, F1-score, AUC-PR (Area Under the Precision-Recall Curve), G-mean, hoặc Matthews Correlation Coefficient (MCC).
3. Cost-Sensitive Learning (Học nhạy cảm với chi phí): Gán trọng số (chi phí) khác nhau cho việc phân loại sai các lớp. Cụ thể là tăng chi phí/hình phạt cho việc phân loại sai lớp thiểu số. Nhiều thuật toán cho phép thiết lập trọng số lớp (ví dụ: Logistic Regression, SVM, Decision Trees).
4. Sử dụng thuật toán phù hợp: Một số thuật toán hoặc biến thể thuật toán hoạt động tốt hơn với dữ liệu mất cân bằng, ví dụ như các phương pháp Ensemble (Balanced Random Forest, EasyEnsemble, RUSBoost).


    c) Giải thích tại sao Accuracy không phải lúc nào cũng là độ đo tốt nhất. Cho ví dụ minh họa. 

Accuracy không phải lúc nào cũng là độ đo tốt nhất, đặc biệt là khi tập dữ liệu bị mất cân bằng hoặc khi chi phí của các loại lỗi là khác nhau.

***Giải thích*** : 
 - Accuracy có thể đưa ra một bức tranh sai lệch về hiệu suất của mô hình trên các tập dữ liệu mất cân bằng. Một mô hình có thể đạt Accuracy rất cao chỉ bằng cách dự đoán thiên về lớp đa số, trong khi hoàn toàn thất bại trong việc xác định lớp thiểu số, mà lớp thiểu số này thường lại là lớp quan trọng hơn (ví dụ: phát hiện bệnh hiếm, giao dịch gian lận).
 - Ví dụ minh họa: Xét bài toán phát hiện gian lận thẻ tín dụng. Giả sử trong 1000 giao dịch, chỉ có 10 giao dịch là gian lận (lớp thiểu số - 1%) và 990 giao dịch là hợp lệ (lớp đa số - 99%).
    + Một mô hình đơn giản luôn dự đoán "không gian lận" cho mọi giao dịch sẽ đạt Accuracy = 990/1000 = 99%.
    + Mặc dù Accuracy rất cao, mô hình này hoàn toàn vô dụng vì nó không phát hiện được bất kỳ giao dịch gian lận nào (Recall cho lớp gian lận = 0/10 = 0%).
    + Trong trường hợp này, việc bỏ sót một giao dịch gian lận (False Negative) thường có hậu quả nghiêm trọng hơn nhiều so với việc cảnh báo nhầm một giao dịch hợp lệ (False Positive). Do đó, các độ đo như Recall, Precision, và F1-score sẽ phản ánh đúng hơn hiệu quả thực sự của mô hình trong việc giải quyết vấn đề.